Update the compile time forcing docs (#2548)
chef/chef-web-docs: layouts/shortcodes/resources_common_compile_begin.md

the specified action during the compile phase. For example:

``` ruby
build_essential 'Install compilers' do
  action :nothing
end.run_action(:install)
```

where `action` is set to `:nothing` to ensure the `run_action` is run
during the compile phase and not later during the execution phase.

This can be simplified by using the `compile_time` flag in Chef Infra
Client 16 and later versions:

``` ruby
build_essential 'Install compilers' do
  compile_time true
end
```

That flag both forces the resource to run at compile time and sets the

> converge action to :nothing.

Reviewer: ```suggestion
converge action to `:nothing`.
```

The following examples show when (and when not) to use `run_action`.

**Using Custom Resources preferred to forcing to compile time**

Compile time execution is often used to install gems before requiring
them in recipe code.

This is a poor pattern, since gems may depend on native gems, which

> may result in needing to install compilers at compile time.

Reviewer: ```suggestion
This is a poor pattern since gems may depend on native gems which
may require installing compilers at compile time.
```

``` ruby
build_essential 'Install compilers' do
  compile_time true
end

chef_gem 'aws-dsk' do
  compile_time true
end

require 'aws-sdk'
```

A better strategy is to move the code which requires the gem into
a custom resource.  Since all the actions of custom resources run
at converge time, this has the effect of deferring the require of

> the gem to later in the overall chef-client execution.  Unified

Reviewer: ```suggestion
A better strategy is to move the code, which requires the gem, into
a custom resource. Since all the actions of custom resources run
at converge time, this defers requiring 
the gem until later in the overall Chef Infra Client execution. Unified
```

mode can also be used in the resource to eliminate compile/converge
mode issues entirely:

``` ruby
unified_mode true

action :run do
  build_essential 'Install compilers'

  chef_gem 'aws-sdk'

  require 'aws-sdk'
end
```

**Download and parse a configuration file**

> A common use case is to download a configuration file, parse it and then

Reviewer: ```suggestion
A common use case is to download a configuration file, parse it, and then
```

use the values in templates and to control other configuration.

> And important distinction to make is that the downloaded configuration file
> is being used internally by the Chef Infra Client as temporary state.

Reviewer: Not sure of the context here but maybe one of these suggestions: 

```suggestion
An important distinction to make is that the downloaded configuration file
only exists in a temporary state to be used by the Chef Infra Client. It will
not be used directly by the system or applications that are managed by the
Chef Infra Client.
```
Or maybe:

`used internally while the Chef Infra Client is in a temporary state.`

Author: definitely not the second one.

Author: So the file that is being downloaded does not really affect the running system or the applications.  It isn't a file in /etc, it isn't some kind of O/S config, it isn't an application config file.  So the file is downloaded, then the recipe reads it, then the chef-client goes and does the REAL work of configuring the system.  So it is internal or temporary or transient state that is only consumed by chef-client itself.

Dunno if that helps.

Also, the file clearly gets downloaded to the filesystem so technically the running state of the node does change.  But the difference I'm trying to highlight is that the change in the state is entirely to a "scratchpad", or that the downloaded file is part of the "preparation" to do actual work.

> To download and parse a JSON file and render it in a template it makes sense

Reviewer: ```suggestion
To download and parse a JSON file and render it in a template, it makes sense
```

> to eagerly download the file in compile time:

Reviewer: ```suggestion
to download the file during compile time:
```

``` ruby
  # the remote_file is being downloaded to a temporary file
  remote_file "#{Chef::Config[:file_cache_path]}/users.json" do
    source "https://jsonplaceholder.typicode.com/users"
    compile_time true
  end

  # this parsing needs to happen after the remote_file is downloaded, but will
  # be executed at compile time.
  array = JSON.parse(IO.read("#{Chef::Config[:file_cache_path]}/users.json")

  # the `array.last["phone"]` expression here will also be evaluated at compile
  # time and must be lazied via wrapping the expresssion in `lazy {}`
  file "/tmp/phone_number.txt" do
    content array.last["phone"]
  end
```

> This is considerably cleaner than the alternative of lazying both the parsing of the

Reviewer: ```suggestion
This is considerably cleaner than the alternative of lazy evaluating both the parsing of the
```

> JSON and the rendering of the data into the file template, which will be caused if

Reviewer: ```suggestion
JSON and the rendering of the data into the file template, which will happen if
```

the `remote_file` resource is not run at compile time:

``` ruby
  # the execution of this is now deferred
  remote_file "#{Chef::Config[:file_cache_path]}/users.json" do
    source "https://jsonplaceholder.typicode.com/users"
  end

  # it is necessary due to lexical scoping issues to create this variable here
  array = nil

  # the parsing of the JSON is now deferred due to the ruby_block
  ruby_block "parse JSON" do
    block do
      array = JSON.parse(IO.read("#{Chef::Config[:file_cache_path]}/users.json")
    end
  end

  # the argument to the content property must now also be deferred
  file "/tmp/phone_number.txt" do
    content lazy { array.last["phone"] }
  end
```

> This is an example of code that overly uses deferred execution, uses more "lazy" evaluation, and is

Reviewer: ```suggestion
This is an example of code that overuses deferred execution, uses more "lazy" evaluation, and is
```

``` ruby
execute 'ifconfig'

package 'vim-enhanced' do
  compile_time true
  notifies :run, 'execute[ifconfig]', :immediately
end
```

A better approach in this type of situation is to install the package
before the resource collection is built to ensure that it is available
to other resources later on.

The best approach to this problem is to use `unified mode` which eliminates
the compile time and converge time distinction, while allowing notifications
to work correctly.

**Resources that are forced to compile time by default**

> The `ohai_hint` and `hostname` resources are forced to compile time by default.

Reviewer: ```suggestion
The `ohai_hint` and `hostname` resources run at compile time by default.
```

This is due to the fact that later resources may consume the node attributes which
are set by those resources leading to excessive use of `lazy` in subsequent
resources (and similar issues to the `remote_file` example above).

> The `chef_gem` resource used to be forced to compile time by default, but that was
> removed in favor of the recommendation to move such code to custom resources.

Reviewer: ```suggestion
The `chef_gem` resource used to execute at compile time by default, but now we
recommend that users move code that executes at compile time to custom resources.
```
